File: euxis-gateway/tests/test_mcp.py
import asyncio
import json
import pytest
from unittest.mock import AsyncMock, MagicMock
from fastapi import WebSocketDisconnect
from gateway.mcp import MCPHost

class DummyState:
    def __init__(self):
        self.started_at = 0
        self.sessions_active = 1

class DummyWS:
    def __init__(self, messages):
        self.messages = messages
        self.sent = []
        self.accept_called = False

    async def accept(self):
        self.accept_called = True

    async def receive_text(self):
        if not self.messages:
            raise WebSocketDisconnect()
        return self.messages.pop(0)

    async def send_text(self, text):
        self.sent.append(json.loads(text))

@pytest.mark.asyncio
async def test_mcp_host_initialization():
    state = DummyState()
    host = MCPHost(state, {})
    assert "get_metrics" in host.tools

@pytest.mark.asyncio
async def test_mcp_handle_connection():
    state = DummyState()
    host = MCPHost(state, {})
    
    messages = [
        json.dumps({"method": "not_initialize", "id": 1}),
        json.dumps({"method": "initialize", "id": 2, "params": {}}),
        json.dumps({"method": "notifications/initialized"}),
        json.dumps({"method": "tools/list", "id": 3}),
        json.dumps({"method": "tools/call", "id": 4, "params": {"name": "get_metrics"}}),
        json.dumps({"method": "tools/call", "id": 5, "params": {"name": "unknown_tool"}}),
        json.dumps({"method": "ping", "id": 6}),
        json.dumps({"method": "unknown_method", "id": 7}),
        "invalid_json"
    ]
    
    ws = DummyWS(messages)
    await host.handle_connection(ws)
    
    assert ws.accept_called
    assert ws.sent[0]["error"]["code"] == -32002  # Not initialized
    assert ws.sent[1]["result"]["serverInfo"]["name"] == "euxis-gateway"  # Initialize
    assert "tools" in ws.sent[2]["result"]  # tools/list
    assert "content" in ws.sent[3]["result"]  # get_metrics call
    assert ws.sent[4]["error"]["code"] == -32601  # Unknown tool
    assert ws.sent[5]["result"] == {}  # Ping
    assert ws.sent[6]["error"]["code"] == -32601  # Unknown method
    assert ws.sent[7]["error"]["code"] == -32700  # Parse error

@pytest.mark.asyncio
async def test_mcp_tool_execution_error():
    state = DummyState()
    host = MCPHost(state, {})
    
    async def bad_handler():
        raise Exception("Tool error")
        
    host.register_tool("bad_tool", "desc", {}, bad_handler)
    
    messages = [
        json.dumps({"method": "initialize", "id": 1}),
        json.dumps({"method": "tools/call", "id": 2, "params": {"name": "bad_tool"}})
    ]
    ws = DummyWS(messages)
    await host.handle_connection(ws)
    
    assert ws.sent[1]["error"]["code"] == -32603

@pytest.mark.asyncio
async def test_mcp_default_tools():
    state = DummyState()
    host = MCPHost(state, {})
    
    metrics = await host._tool_get_metrics("sess")
    assert "uptime_ms" in metrics
    
    sig = await host._tool_sign_payload("data", "key")
    assert "signature" in sig

File: cmd/gateway/tests/test_mcp.cpp
#include <gtest/gtest.h>
#include <sodium.h>

#include "euxis/gateway/mcp.hpp"

namespace euxis::gateway {
namespace {

class McpHostTest : public ::testing::Test {
protected:
    static void SetUpTestSuite() {
        ASSERT_GE(sodium_init(), 0);
    }

    McpHost host;

    static auto make_request(const std::string& method,
                             const nlohmann::json& params = {},
                             const nlohmann::json& id = 1) -> nlohmann::json {
        nlohmann::json req = {{"jsonrpc", "2.0"}, {"method", method}, {"id", id}};
        if (!params.is_null()) {
            req["params"] = params;
        }
        return req;
    }

    void initialize() {
        host.handle_request(make_request("initialize"));
    }
};

TEST_F(McpHostTest, InitializationRequired) {
    auto res = host.handle_request(make_request("tools/list"));
    EXPECT_EQ(res["error"]["code"], -32002);
    EXPECT_EQ(res["error"]["message"], "Server not initialized");
}

TEST_F(McpHostTest, InitializeReturnsCapabilities) {
    auto res = host.handle_request(make_request("initialize"));
    ASSERT_TRUE(res.contains("result"));
    auto& result = res["result"];
    EXPECT_EQ(result["protocolVersion"], "2024-11-05");
    EXPECT_EQ(result["serverInfo"]["name"], "euxis-gateway");
    EXPECT_TRUE(result["capabilities"].contains("tools"));
}

TEST_F(McpHostTest, ToolsListReturnsRegisteredTools) {
    initialize();
    auto res = host.handle_request(make_request("tools/list"));
    auto& tools = res["result"]["tools"];
    ASSERT_GE(tools.size(), 2);

    bool has_metrics = false;
    bool has_sign = false;
    for (const auto& t : tools) {
        if (t["name"] == "get_metrics") has_metrics = true;
        if (t["name"] == "sign_payload") has_sign = true;
    }
    EXPECT_TRUE(has_metrics);
    EXPECT_TRUE(has_sign);
}

TEST_F(McpHostTest, ToolsCallGetMetrics) {
    initialize();
    nlohmann::json params = {{"name", "get_metrics"}, {"arguments", {}}};
    auto res = host.handle_request(make_request("tools/call", params));
    ASSERT_TRUE(res.contains("result"));
    auto& content = res["result"]["content"];
    ASSERT_EQ(content.size(), 1);
    EXPECT_EQ(content[0]["type"], "text");

    auto text = nlohmann::json::parse(content[0]["text"].get<std::string>());
    EXPECT_TRUE(text.contains("uptime_ms"));
    EXPECT_TRUE(text.contains("sessions_active"));
    EXPECT_TRUE(text.contains("timestamp"));
}

TEST_F(McpHostTest, ToolsCallSignPayload) {
    initialize();
    nlohmann::json params = {
        {"name", "sign_payload"},
        {"arguments", {{"payload", "test data"}}}};
    auto res = host.handle_request(make_request("tools/call", params));
    ASSERT_TRUE(res.contains("result"));
    auto& content = res["result"]["content"];
    ASSERT_EQ(content.size(), 1);

    auto text = nlohmann::json::parse(content[0]["text"].get<std::string>());
    EXPECT_EQ(text["payload"], "test data");
    EXPECT_EQ(text["algorithm"], "ed25519");
    EXPECT_FALSE(text["signature"].get<std::string>().empty());
}

TEST_F(McpHostTest, ToolsCallUnknownTool) {
    initialize();
    nlohmann::json params = {{"name", "nonexistent"}, {"arguments", {}}};
    auto res = host.handle_request(make_request("tools/call", params));
    EXPECT_EQ(res["error"]["code"], -32601);
}

TEST_F(McpHostTest, PingReturnsEmptyResult) {
    initialize();
    auto res = host.handle_request(make_request("ping"));
    ASSERT_TRUE(res.contains("result"));
    EXPECT_TRUE(res["result"].is_object());
    EXPECT_TRUE(res["result"].empty());
}

TEST_F(McpHostTest, CustomToolRegistration) {
    initialize();
    host.register_tool("echo", "Echo arguments back",
                       {{"type", "object"}},
                       [](const nlohmann::json& args) { return args; });

    nlohmann::json params = {
        {"name", "echo"},
        {"arguments", {{"msg", "hello"}}}};
    auto res = host.handle_request(make_request("tools/call", params));
    auto text = nlohmann::json::parse(
        res["result"]["content"][0]["text"].get<std::string>());
    EXPECT_EQ(text["msg"], "hello");
}

TEST_F(McpHostTest, ToolHandlerException) {
    initialize();
    host.register_tool("fail", "Always fails",
                       {{"type", "object"}},
                       [](const nlohmann::json&) -> nlohmann::json {
                           throw std::runtime_error("intentional");
                       });

    nlohmann::json params = {{"name", "fail"}, {"arguments", {}}};
    auto res = host.handle_request(make_request("tools/call", params));
    EXPECT_EQ(res["error"]["code"], -32603);
}

TEST_F(McpHostTest, ParseErrorOnInvalidRequest) {
    // Missing jsonrpc field
    nlohmann::json bad = {{"method", "ping"}};
    auto res = host.handle_request(bad);
    EXPECT_EQ(res["error"]["code"], -32600);
}

TEST_F(McpHostTest, UnknownMethod) {
    initialize();
    auto res = host.handle_request(make_request("nonexistent/method"));
    EXPECT_EQ(res["error"]["code"], -32601);
}

TEST_F(McpHostTest, NotificationsInitializedAck) {
    auto res = host.handle_request(make_request("notifications/initialized"));
    ASSERT_TRUE(res.contains("result"));
}

// --- Coverage: mcp.cpp lines 77-78 (tools() const accessor) ---
TEST_F(McpHostTest, ToolsAccessorReturnsList) {
    const auto& tools = host.tools();
    ASSERT_GE(tools.size(), 2u);
    bool has_metrics = false;
    bool has_sign = false;
    for (const auto& t : tools) {
        if (t.name == "get_metrics") has_metrics = true;
        if (t.name == "sign_payload") has_sign = true;
    }
    EXPECT_TRUE(has_metrics);
    EXPECT_TRUE(has_sign);
}

// --- Coverage: mcp.cpp line 77 (tools() on host with custom tool) ---
TEST_F(McpHostTest, ToolsAccessorIncludesCustomTools) {
    host.register_tool("custom", "Custom tool", {{"type", "object"}},
                       [](const nlohmann::json& args) { return args; });
    const auto& tools = host.tools();
    ASSERT_GE(tools.size(), 3u);
    bool has_custom = false;
    for (const auto& t : tools) {
        if (t.name == "custom") has_custom = true;
    }
    EXPECT_TRUE(has_custom);
}

} // namespace
} // namespace euxis::gateway
